sql/sql_connect.cc

```diff
@@ -52,13 +52,97 @@ extern mysql_mutex_t LOCK_global_user_client_stats;
 extern mysql_mutex_t LOCK_global_table_stats;
 extern mysql_mutex_t LOCK_global_index_stats;
 extern vio_keepalive_opts opt_vio_keepalive;
+extern mysql_mutex_t LOCK_user_failed_login;
+
+/* Delay user login response mutex and lock */
+PSI_mutex_key key_connection_delay_mutex= PSI_NOT_INSTRUMENTED;
+PSI_cond_key key_connection_delay_wait= PSI_NOT_INSTRUMENTED;
+PSI_stage_info stage_waiting_in_user_login_failed_delay= {
+    0, "Waiting in connection_control stage", 0};
+
 
 /*
   Get structure for logging connection data for the current user
 */
 
 #ifndef NO_EMBEDDED_ACCESS_CHECKS
 static HASH hash_user_connections;
+HASH user_login_failed_hash;
```

Comment:
On the other hand, do you need a hash at all? failed password attempts are counted in `ACL_USER` already, you can simply use that.

Reply:
Yes, use password errors in ACL_USER is very simple, and don't to introduce additional HASH and less lock . But here maybe one problem , ACL_USER is count failed password attempts by acl user **not connect user/host** . In this case , acl user is app@% , if someone login as  app user with many wrong times  , it will delay  **application  server in production connect**. I don't know  this is acceptable.

Reply:
I think it is. I'd say it's important to behave consistently in all cases of "wrong password" reactions. If the whole ACL_USER is locked, then the timeout should also apply to the ACL_USER.

Supposedly, it was done to prevent distributed password brute-forcing from many different IPs.